Entry points for recurring jobs

br648 opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR seeks to allow fine grain control over the jobs within the OTP middleware which should be started. This will then allow various instances of OTP middleware to be started which focus on specific jobs instead of all jobs running under a single instance.

Breaking Change!

**_--endpoints-only_** or **_-E_** must be defined in order to get endpoints.

br648 committed:
feat(Command line control over LB and recurring jobs): Update to define via the command line which j

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/utils/CommandLineProcessorTest.java:
```diff
@@ -0,0 +1,115 @@
+            return this;
+        }
+
+        public CommandLineTestCase withCommand(String[] command) {
```
Is it possible to use `String... command` to simplify the syntax of the test cases?

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/utils/CommandLineProcessorTest.java:
```diff
@@ -0,0 +1,115 @@
+            return this;
+        }
+
+        public CommandLineTestCase withRecurringJob(Set<RecurringJob> recurringJobs) {
```
Rename to `withRecurringJobs`.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/CommandLineProcessor.java:
```diff
@@ -0,0 +1,117 @@
+            if (command.equalsIgnoreCase("-recurringjobs")) {
+                defineRecurringJobs(commandValues);
+            }
+            if (command.equalsIgnoreCase("-loadbalance")) {
```
Trying to find a better name for this option without having to reference load balancers... How about something like `--endpoints-only` or `--api-endpoints` because this serves the API endpoints only, correct? (so that a load balancer can dispatch web requests).

binh-dam-ibigroup replied:
For command line flags, we should probably follow the convention used by many Unix programs - typically  double-dash + kebab-case-command (e.g. `--load-balancer`) and either one dash + one letter/number (e.g. `-L`) for shorthand syntax. (Note: OTP uses pascalCase commands as in `--loadBalancer` and that's ok if we want to adopt that.)

br648 replied:
My thinking was to keep this generic, so anything that could be load balanced would be covered. At the moment, the only item is the API, so will change it to suit this.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/recurringjobs/RecurringJob.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public enum RecurringJob {
```
This is probably okay as first approach. I had in mind that some recurring jobs should eventually be handled by the OS (e.g. cron) so that they can be started/terminated regardless of completion status.

br648 replied:
This will at least allow individual jobs to be stopped/started.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/ConfigUtils.java:
```diff
@@ -70,17 +70,17 @@ public static boolean getBooleanEnvVar(String var) {
         } else {
-            LOG.info("Loading env.yml: {}", args[0]);
-            envConfig = yamlMapper.readTree(new FileInputStream(args[0]));
+            LOG.info("Loading env.yml: {}", configFile);
```
Nit - Could you combine these two blocks?

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/OtpMiddlewareMain.java:
```diff
@@ -57,58 +54,35 @@ public class OtpMiddlewareMain {
+            if (!recurringJobs.isEmpty()) {
+                recurringJobs.forEach(RecurringJob::schedule);
+                try {
+                    // Block the main thread until the scheduler is terminated to prevent the application from
```
Maybe this is why I'd prefer recurring jobs to be scheduled by the OS itself?

br648 replied:
Yes, this isn't ideal. My approach for this first iteration was to not change the existing scheduled services. In the case of `BugsnagJobs` it actually schedules three different jobs with different parameters. Perhaps for a specific job (e.g. --cron-job) that qualifies, I can rearchitecture to have a `Main` entry point? We still need to consider backwards compatibility.

binh-dam-ibigroup replied:
It's fine to take it one step at a time.

binh-dam-ibigroup reviewed (commented):
First round of comments. I still need to play around with the changes, but I would like to explore using cron for some recurring jobs.

br648 committed:
improvement(Addressed PR feedback):

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/utils/CommandLineProcessorTest.java:
```diff
@@ -0,0 +1,122 @@
+            return this;
+        }
+
+        public CommandLineTestCase withLoadBalanced(boolean isLoadBalanced) {
```
Rename to `withEndpoints` (rename the parameter accordingly too).

br648 replied:
Doh! My bad.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/utils/CommandLineProcessorTest.java:
```diff
@@ -0,0 +1,122 @@
+    private static class CommandLineTestCase {
+        public String configFile = DEFAULT_ENV;
+        public Set<RecurringJob> recurringJobs = RecurringJob.getAllRecurringJobs();
+        public boolean isLoadBalanced = true;
```
Rename to `hasEndpoints`.

br648 replied:
Doh! My bad.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/OtpMiddlewareMain.java:
```diff
@@ -57,58 +54,35 @@ public class OtpMiddlewareMain {
             }
         }
+
+        if (processor.hasEndPoints()) {
```
Move this block to before the scheduler stuff. Otherwise, because of the thread blocking above, this block is never executed (until termination), and therefore no endpoints are available.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/CommandLineProcessor.java:
```diff
@@ -0,0 +1,120 @@
+     * Define if the end points should be enabled.
+     */
+    private void defineEndPoints(Set<String> endPointArguments) {
+        if (endPointArguments.size() == 1 && endPointArguments.contains("yes")) {
```
In how the parsing code works, is it possible to just pass a `--endpoint-only` flag without the 
"yes" or "no"? (so that `--endpoints-only` without a parameter implies "yes")

br648 replied:
It can and I have updated, but this is now NOT backward compatible.

binh-dam-ibigroup replied:
With the latest changes, `--endpoints no` exposes endpoints.
I should have clarified we need the following:
- `--endpoints yes` => expose endpoints
- `--endpoints no` => hide endpoints
- `--endpoints` (for convenience) => same as `--endpoints yes`.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/CommandLineProcessor.java:
```diff
@@ -0,0 +1,120 @@
+ * /configurations/test/env.yml --endpoints-only yes --recurring-jobs monitor-all-trips-job
+ */
+public class CommandLineProcessor {
+    public static final List<String> END_POINTS_ONLY_FLAGS = List.of("--endpoints-only", "-E");
```
Sorry, rename to `--endpoints` because the flag is not meant to exclude the other features.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/CommandLineProcessor.java:
```diff
@@ -0,0 +1,120 @@
+        }
+        Map<String, Set<String>> groupedParams = parseCommandLineArguments(arguments);
+        groupedParams.forEach((flag, commandValues) -> {
+            if (RECURRING_JOB_FLAGS.contains(flag)) {
+                defineRecurringJobs(commandValues);
+            }
```
To only enable the endpoints and run no recurring jobs, the command line would read `--recurring-jobs`, and that feels kinda strange. Maybe including a `--recurring-jobs no` or `--recurring-jobs none` in addition to just `--recurring-jobs` would help reduce confusion.

br648 replied:
I have updated with `--recurring-jobs none`

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/CommandLineProcessor.java:
```diff
@@ -0,0 +1,120 @@
+
+    /**
+     * Parse the arguments provided on the command line. To maintain backward compatibility, if a config file is
+     * required, it must be the first argument and no command e.g. -config is NOT required.
```
```suggestion
     * required, it must be the first argument, and no flag (e.g. --config) is used.
```

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/recurringjobs/RecurringJob.java:
```diff
@@ -0,0 +1,64 @@
+        this.recurringJobScheduler = recurringJobScheduler;
+    }
+
+    public static RecurringJob getJobFromCommandLineArgument(String commandLineArgument) {
+        return Arrays.stream(RecurringJob.values())
+            .filter(job -> job.commandLineName.equalsIgnoreCase(commandLineArgument))
+            .findFirst()
+            .orElse(null);
+    }
```
How does this method differ from the native `Enum.valueOf()`?

br648 replied:
It is similar, but the IllegalArgumentException I throw contains more relevant information i.e. valid jobs.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/recurringjobs/RecurringJob.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public static Set<RecurringJob> getAllRecurringJobs() {
+        return new HashSet<>(Arrays.asList(RecurringJob.values()));
```
```suggestion
        return Set.of(RecurringJob.values());
```

br648 replied:
The Set needs to be modifiable.

binh-dam-ibigroup replied:
> The Set needs to be modifiable.

Please add a comment about where this requirement comes from.

binh-dam-ibigroup reviewed (changes requested):
Moving in the right direction, but endpoints are not working at all as written. See also other suggestions.

br648 committed:
improvement(Address PR feedback): Made changes related to command line params

BREAKING CHANGE: The command --endpoints (or -E) is required for end points to be initialised.

br648 commented:
@binh-dam-ibigroup because of this:

BREAKING CHANGE: The command --endpoints (or -E) is required for end points to be initialised.

CI is failing. Is this something you can address?

br648 committed:
improvement(Updates to address default command line args): Updated unit tests to include end points

br648 committed:
improvement(OtpMiddlewareTestEnvironment.java): Fixed file test

br648 committed:
improvement(README.md): Updated example to include end points which is now required.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/OtpMiddlewareMain.java:
```diff
@@ -57,58 +54,37 @@ public class OtpMiddlewareMain {
-        initializeHttpEndpoints();
+        if (processor.hasEndPoints()) {
+            // This must take place before the main thread is blocked below.
+            TemplateUtils.initialize();
```
`TemplateUtils.initialize();` is for notification content... Move this to outside the if block. It is needed for endpoint mode (when saving a trip the first time, monitored trip stuff runs and notifications might be sent then) and in job-only mode. Without that, in jobs-only mode, notifications will not work.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/OtpMiddlewareMain.java:
```diff
@@ -57,58 +54,37 @@ public class OtpMiddlewareMain {
+                }
             }
         }
+
```
Remove extra blank line

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/CommandLineProcessor.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public static String[] getDefaultArguments() {
+        return new String[]{END_POINTS_ONLY_FLAGS.get(0)};
+    }
+
+    public static String[] getDefaultArguments(String configFile) {
+        return new String[]{configFile, END_POINTS_ONLY_FLAGS.get(0)};
```
Insert space after `]` (2 instances).

binh-dam-ibigroup reviewed (changes requested):
Sorry, a few more chinks:
- Make `--endpoints no` not expose endpoints
- Initialize templates for all execution modes.

br648 committed:
improvement(Addressed PR feedback): Explict updates around defining end points

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/CommandLineProcessor.java:
```diff
@@ -0,0 +1,129 @@
+            hadEndPoints = true;
+        } else if (endPointArguments.size() == 1 && endPointArguments.contains("no")) {
+            hadEndPoints = false;
+        }
```
Nice to have: I think we need to throw an `IllegalArgumentException` in other cases.

binh-dam-ibigroup reviewed (approved):
One nice-to-have but o/w I think it looks good with all the changes.

br648 committed:
improvement(Address PR feedback): Throw illigal argument exception for unknown end point only comman

JymDyerIBI reviewed (approved):
👍  Looks good. I'm adapting one of my tickets to use the `RecurringJob` class.
